File: php/src/DigestTokens.php
<?php

namespace EdgeAuth;

class DigestTokens
{
    const digestTokenPrefix = 'DIGEST:';
    const digestAlgorithm = 'SHA512';

    function __construct()
    {
    }

    /**
     * Check if a value is a valid digest token
     * @param string $encodedToken an encoded token
     * @return bool true if the encodedToken is a valid digest token
     */
    public function isDigestToken($encodedToken)
    {
        return isset($encodedToken)
            && is_string($encodedToken)
            && substr($encodedToken, 0, strlen(self::digestTokenPrefix)) == self::digestTokenPrefix;
    }

    /**
     * Verifies and decodes a digest token.
     * @param string $secret the shared secret used to sign the token
     * @param string $encodedToken encodedToken the encoded token
     * @return mixed|object  An object {verified,code,value} with verified set to true if the token was successfully verified. And with verified set to false if the token was not verified. In that case, code is provided to indicate the type of problem.
     * @throws \Exception
     */
    public function verifyAndDecode($secret, $encodedToken)
    {
        if (is_string($secret) === false) {
            throw new \Exception('Secret must be a string.');
        }

        if (is_string($encodedToken) === false) {
            throw new \Exception('Encoded token must be a string.');
        }

        if ($this->isDigestToken($encodedToken) === false) {
            return (object)[
                'verified' => false,
                'code' => 'not-a-digest-token'
            ];
        }

        $encodedDigestToken = substr($encodedToken, strlen(self::digestTokenPrefix));
        $decodedDigestTokenAsString = base64_decode($encodedDigestToken);

        $info = json_decode($decodedDigestTokenAsString);
        if ($info === false) {
            return (object)[
                'verified' => false,
                'code' => 'bad-token'
            ];
        }

        if (!isset($info->applicationId) || !is_string($info->applicationId)) {
            return (object)[
                'verified' => false,
                'code' => 'bad-token'
            ];
        }


        if (!isset($info->digest) || !is_string($info->digest)) {
            return (object)[
                'verified' => false,
                'code' => 'bad-token'
            ];
        }

        if (!isset($info->token) || !is_string($info->token)) {
            return (object)[
                'verified' => false,
                'code' => 'bad-token'
            ];
        }

        $digestAsString = $this->calculateDigest($info->applicationId, $secret, $info->token);
        $digest = $info->digest;

        if ($digestAsString !== $digest) {
            return (object)[
                'verified' => false,
                'code' => 'bad-digest'
            ];
        }

        $value = json_decode($info->token);
        $value->applicationId = $info->applicationId;

        return (object)[
            'verified' => true,
            'code' => 'verified',
            'value' => $value
        ];
    }

    /**
     * Signs and encodes a digest token.
     * @param string $applicationId the application ID used to sign the token
     * @param string $secret the shared secret used to sign the token
     * @param object $token the raw token object to sign
     * @return string
     * @throws \Exception
     */
    public function signAndEncode($applicationId, $secret, $token)
    {
        if (!isset($applicationId) || !is_string($applicationId)) {
            throw new \Exception('Application ID must be a string.');
        }

        if (!isset($secret) || !is_string($secret)) {
            throw new \Exception('Secret must be a string.');
        }

        if (!isset($token) || !is_object($token)) {
            throw new \Exception('Encoded token must be a object.');
        }

        if (!isset($token->expires) || !is_numeric($token->expires)) {
            throw new \Exception('Token must have an expiration (milliseconds since UNIX epoch)');
        }

        if (isset($token->applicationId)) {
            throw new \Exception('Token should not have an application ID property');
        }

        $tokenAsString = json_encode($token);
        $digest = $this->calculateDigest($applicationId, $secret, $tokenAsString);

        $info = (object)[
            'applicationId' => $applicationId,
            'digest' => $digest,
            'token' => $tokenAsString
        ];

        $decodedDigestTokenAsString = json_encode($info, JSON_UNESCAPED_SLASHES);
        $encodedDigestToken = base64_encode($decodedDigestTokenAsString);


        return self::digestTokenPrefix . $encodedDigestToken;
    }

    /**
     * Calculates the digest for a token.
     * @param $applicationId
     * @param $secret
     * @param $token
     * @return string The Base64 encoded digest
     * @throws \Exception
     */
    public function calculateDigest($applicationId, $secret, $token)
    {
        if (!isset($applicationId) || !is_string($applicationId)) {
            throw new \Exception('Application ID must be a string');
        }

        if (!isset($secret) || !is_string($secret)) {
            throw new \Exception('Secret must be a string');
        }

        if (!isset($token) || !is_string($token)) {
            throw new \Exception('Token must be a string');
        }

        // The hmac salt is the concatenation of application ID and secret to eliminate the use of lookup table for brute force attacks.
        $salt = $applicationId . $secret;
        return base64_encode(hash_hmac(self::digestAlgorithm, $token, $salt, true));
    }

}

File: golang/digest_tokens.go
package edgeauth

import (
	"crypto/hmac"
	"crypto/sha512"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
)

type digestInfo struct {
	ApplicationID string `json:"applicationId"`
	Digest        string `json:"digest"`
	Token         string `json:"token"`
}

// DigestTokenPrefix is the prefix used for all digest tokens.
const DigestTokenPrefix = "DIGEST:"

// Check if a value is a valid digest token.
//
// Arguments:
//   encodedToken -- an encoded token
func isDigestToken(encodedToken string) bool {
	return strings.HasPrefix(encodedToken, DigestTokenPrefix)
}

// SignAndEncode signs and encodes a digest token.
//
// Arguments:
//   applicationID -- the application ID used to sign the token
//   secret -- the shared secret used to sign the token
//   token -- the raw token object to sign
func SignAndEncode(applicationID string, secret string, token *Token) (*string, error) {
	if _, ok := token.Get(ExpiresField); !ok {
		return nil, errors.New("Token must have an expiration (milliseconds since UNIX epoch)")
	}
	if _, ok := token.Get(ApplicationIDField); ok {
		return nil, errors.New("Token should not have an ApplicationID property")
	}

	tokenAsString, err := json.Marshal(token)
	if err != nil {
		return nil, err
	}

	digest := calculateDigest(applicationID, secret, tokenAsString)

	info := &digestInfo{
		ApplicationID: applicationID,
		Digest:        digest,
		Token:         string(tokenAsString),
	}

	decodedDigestTokenAsString, err := json.Marshal(info)
	if err != nil {
		return nil, err
	}
	encodedDigestToken := base64.StdEncoding.EncodeToString(decodedDigestTokenAsString)

	retval := DigestTokenPrefix + encodedDigestToken
	return &retval, nil
}

// VerifyAndDecodeResult is the result of verifying and decoding a
// digest token.
type VerifyAndDecodeResult struct {
	Verified bool
	Code     string
	Message  string
	Value    *Token
}

// VerifyAndDecode verifies and decodes a digest token.
//
// Arguments:
//  secret -- the shared secret used to sign the token
//  encodedToken -- the encoded token
func VerifyAndDecode(secret string, encodedToken string) *VerifyAndDecodeResult {
	if !isDigestToken(encodedToken) {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "not-a-digest-token",
		}
	}

	encodedDigestToken := encodedToken[len(DigestTokenPrefix):]
	decodedDigestTokenAsString, err := base64.StdEncoding.DecodeString(encodedDigestToken)
	if err != nil {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "bad-token",
		}
	}

	info := digestInfo{}

	err = json.Unmarshal(decodedDigestTokenAsString, &info)
	if err != nil {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "bad-token",
		}
	}

	if len(info.ApplicationID) == 0 {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "bad-token",
		}
	}
	if len(info.Digest) == 0 {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "bad-token",
		}
	}
	if len(info.Token) == 0 {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "bad-token",
		}
	}

	digestAsString := calculateDigest(info.ApplicationID, secret, []byte(info.Token))
	digest := info.Digest

	if digestAsString != digest {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "bad-digest",
		}
	}

	value := NewToken()
	err = json.Unmarshal([]byte(info.Token), &value)
	if err != nil {
		return &VerifyAndDecodeResult{
			Verified: false,
			Code:     "server-error",
			Message:  err.Error(),
		}
	}

	value.Set(ApplicationIDField, info.ApplicationID)

	return &VerifyAndDecodeResult{
		Verified: true,
		Code:     "verified",
		Value:    value,
	}
}

// Calculates the digest for a token.
//
// Arguments:
//  applicationID -- the application ID used to sign the token
//  secret -- the shared secret used to sign the token
//  token -- encoded token
func calculateDigest(applicationID string, secret string, token []byte) string {
	salt := []byte(applicationID + secret)
	verify := hmac.New(sha512.New, salt)
	verify.Write(token)

	digest := base64.StdEncoding.EncodeToString(verify.Sum(nil))

	return digest
}
